test/proxy_recover_funds.js

```diff
@@ -61,158 +62,180 @@ contract('Proxy funds', accounts => {
-          async () => await recoverEth(target, vault)
-        ))
-
-        it('recovers tokens', async () => {
-          await recoverTokens(target, vault)
-        })
-
-        it('fails if vault is not contract', async () => {
-          await failWithoutVault(target, vault)
-        })
-      })
-
-      context('> Conditional fund recovery', async () => {
-        let conditionalRecoveryAppCode
-
-        before(async () => {
-          conditionalRecoveryAppCode = await AppStubConditionalRecovery.new()
-        })
-
-        beforeEach(async () => {
-          // upgrade app to stub with conditional recovery code
-          await kernel.setApp(APP_BASES_NAMESPACE, appId, conditionalRecoveryAppCode.address)
-          target = AppStub.at(appProxy.address)
-        })
-
-        it('does not allow recovering ETH', skipCoverageIfVaultProxy(
-          // conditional stub doesnt allow eth recoveries
-          () => assertRevert(
-            () => recoverEth(target, vault)
-          )
-        ))
-
-        it('allows recovering tokens', async () => {
-          await recoverTokens(target, vault)
-        })
-      })
-
-      context('> Kernel without proxy', async () => {
-        beforeEach(() => {
-          target = kernel
-        })
-
-        it('recovers ETH', skipCoverageIfVaultProxy(async () =>
-          await recoverEth(target, vault)
-        ))
-
-        it('recovers tokens', async () => {
-          await recoverTokens(target, vault)
-        })
+      beforeEach(async () => {
+        if (kernelType === 'Kernel') {
+          kernel = await Kernel.new(false) // don't petrify so it can be used
+        } else if (kernelType === 'KernelProxy') {
+          kernel = Kernel.at((await KernelProxy.new(kernelBase.address)).address)
+        }
 
-        it('fails if vault is not contract', async () => {
-          await failWithoutVault(target, vault)
-        })
+        await kernel.initialize(aclBase.address, permissionsRoot);
+        const acl = ACL.at(await kernel.acl())
+        const r = await kernel.APP_MANAGER_ROLE()
+        await acl.createPermission(permissionsRoot, kernel.address, r, permissionsRoot)
       })
 
-      context('> Kernel Proxy', async () => {
-        beforeEach(() => {
-          target = Kernel.at(kernelProxy.address)
-        })
-
-        it('recovers ETH', skipCoverageIfVaultProxy(async () =>
-          await recoverEth(target, vault)
-        ))
-
-        it('recovers tokens', async () => {
-          await recoverTokens(target, vault)
-        })
+      // Test both the Vault itself and when it's behind a proxy to make sure their recovery behaviours are the same
+      for (const vaultType of ['Vault', 'VaultProxy']) {
+        const skipCoverageIfVaultProxy = test => {
+          // The VaultMock isn't instrumented during coverage, but the AppProxy is, and so transferring
+          // to the fallback fails when we're testing with the proxy
```

Comment:
We should probably add a comment with the reason why it fails when running the coverage (`.transfer()` hardcoded 2.3k gas limit in Solidity and coverage gas costs due to injected logs).